script/color-contrast.ts

```diff
@@ -0,0 +1,173 @@
+// @ts-ignore
+import { ContrastRequirement, contrastRequirements, cavnasColors } from './color-contrast.config'
+import { Table } from "console-table-printer"
+import { flattenObject } from './lib/flattenObject'
+import colors from "../dist/ts"
+import * as fs from 'fs';
+import { normal } from 'color-blend'
+import { getContrast, parseToRgba, rgba } from 'color2k'
+/**
+ * Type definitions
+ */
+type contrastTestResult = {
+  contrastPair: string;
+  pass: string;
+  contrastRatio: string;
+  minimumContrastRatio: string;
+}
+/**
+ * getOpaqueColor
+ * @description calculates the rgb string without opacity from a color with opacity and the background it is palced on
+ * @param color 
+ * @param background 
+ * @returns rgb string
+ */
+const getOpaqueColor = (color: string, background: string): string => {
+  const [colorR, colorG, colorB, colorAlpha] = parseToRgba(color)
+  // color is not transparent
+  if (colorAlpha === 1) {
+    return color
+  }
+  // parse BG
+  const [bgR, bgG, bgB, bgAlpha] = parseToRgba(background)
+  // return mixed color
+  const mixed = normal({ r: bgR, g: bgG, b: bgB, a: bgAlpha }, { r: colorR, g: colorG, b: colorB, a: colorAlpha })
+  return rgba(mixed.r, mixed.g, mixed.b, mixed.a)
+}
+/**
+ * runContrastTest
+ * @description runs through all color pairs of a theme and checks the contrasts
+ * @param colorPairs 
+ * @param colors 
+ * @returns contrastTestResult
+ */
+const runContrastTest = (colorPairs: ContrastRequirement[], colors: any): contrastTestResult[] =>
+  // Object.fromEntries(
+  colorPairs.flatMap(([minimumContrast, colorA, colorB, options]: ContrastRequirement) => {
+    // concat name
+    const contrastPair = `${colorA} vs. ${colorB}`
+    // build required string
+    const minimumContrastRatio = `${minimumContrast}:1`;
+    // colorB is fully opaque
+    if (parseToRgba(colors[colorB])[3] === 1) {
+      return {
+        contrastPair,
+        ...testContrast(minimumContrast, colors[colorA], colors[colorB], undefined, contrastPair),
+        minimumContrastRatio
+      }
+    }
+    // if colorB is semi-transparent 
+    // get the correct canvas colors to test agains
+    let canvasColorArrays = cavnasColors
+    // overwrite background if custom canvas array is set
+    if (options?.canvas) canvasColorArrays = options.canvas
+    // test transparent colorB with default bgs `cavnasColors`
+    return canvasColorArrays.map(bg => ({
+      contrastPair: `${contrastPair} on ${bg}`,
+      ...testContrast(minimumContrast, colors[colorA], colors[colorB], colors[bg], contrastPair),
+      minimumContrastRatio
+    })
+    )
+  })
+// )
+/**
+ * testContrast
+ * @description test the contrast of two colors against each other
+ * @param minimumContrast 
+ * @param colorA 
```

Comment:
We could, but this would not be truthful for testing e.g. default text color vs. link color. I had fg vs. bg first.

Happy to adjust it again, what do you think?